packages/hardhat-core/src/internal/solidity/compiler/downloader.ts

```diff
@@ -24,301 +19,337 @@ export enum CompilerPlatform {
   WASM = "wasm",
 }
 
-interface CompilerPath {
-  compilerPath: string; // absolute path
+export interface Compiler {
+  version: string;
+  longVersion: string;
+  compilerPath: string;
+  isSolcJs: boolean;
+}
+
+interface CompilerBuild {
+  path: string;
+  version: string;
+  build: string;
+  longVersion: string;
+  keccak256: string;
+  urls: string[];
   platform: CompilerPlatform;
 }
 
-export interface CompilersList {
+interface CompilerList {
   builds: CompilerBuild[];
-  releases: {
-    [version: string]: string;
-  };
+  releases: { [version: string]: string };
   latestRelease: string;
 }
 
-const log = debug("hardhat:core:solidity:downloader");
-
-const COMPILER_FILES_DIR_URL_SOLC = "https://binaries.soliditylang.org/";
-
-async function downloadFile(
-  url: string,
-  destinationFile: string
-): Promise<void> {
-  const { download } = await import("../../util/download");
-  log(`Downloading from ${url} to ${destinationFile}`);
-  await download(url, destinationFile);
+/**
+ * A compiler downloader which must be specialized per-platform. It can't and
+ * shouldn't support multiple platforms at the same time.
+ */
+export interface ICompilerDownloader {
+  /**
+   * Returns true if the compiler has been downloaded.
+   *
+   * This function access the filesystem, but doesn't modify it.
+   */
+  isCompilerDownloaded(version: string): Promise<boolean>;
+
+  /**
+   * Downloads the compiler for a given version, which can later be obtained
+   * with getCompiler.
+   */
+  downloadCompiler(version: string): Promise<void>;
+
+  /**
+   * Returns the compiler, which MUST be downloaded before calling this function.
+   *
+   * Returns undefined if the compiler has been downloaded but can't be run.
+   *
+   * This function access the filesystem, but doesn't modify it.
+   */
+  getCompiler(version: string): Promise<Compiler | undefined>;
 }
 
-type CompilerDownloaderOptions = Partial<{
-  download: (url: string, destinationFile: string) => Promise<void>;
-  forceSolcJs: boolean;
-}>;
+/**
+ * Default implementation of ICompilerDownloader.
+ *
+ * Important things to note:
+ *   1. If a compiler version is not found, this downloader may fail.
+ *    1.1. It only re-downloads the list of compilers once every X time.
+ *      1.1.1 If a user tries to download a new compiler after X time has passed
```

Comment:
Shouldn't this be _before_ X time has passed?

Reply:
right